ops/testing.py

```diff
@@ -242,25 +233,36 @@ def update_relation_data(self, relation_id, app_or_unit, key_values):
         :param key_values: Each key/value will be updated in the relation data.
         :return: None
         """
+        relation_name = self._backend._relation_names[relation_id]
+        relation = self._model.get_relation(relation_name, relation_id)
+        if '/' in app_or_unit:
+            entity = self._model.get_unit(app_or_unit)
+        else:
+            entity = self._model.get_app(app_or_unit)
+        rel_data = relation.data.get(entity, None)
+        if rel_data is not None:
+            # If we have read and cached this data, make sure we invalidate it
+            rel_data._invalidate()
+
+        is_peer = self._meta.relations[relation_name].role == 'peers'
         new_values = self._backend._relation_data[relation_id][app_or_unit].copy()
         for k, v in key_values.items():
             if v == '':
                 new_values.pop(k, None)
             else:
                 new_values[k] = v
         self._backend._relation_data[relation_id][app_or_unit] = new_values
-        relation_name = self._backend._relation_names[relation_id]
-        if self._model is not None:
-            relation = self._model.get_relation(relation_name, relation_id)
-            if '/' in app_or_unit:
-                entity = self._model.get_unit(app_or_unit)
-            else:
-                entity = self._model.get_app(app_or_unit)
-            rel_data = relation.data.get(entity, None)
-            if rel_data is not None:
-                # If we have read and cached this data, make sure we invalidate it
-                rel_data._invalidate()
-        # TODO: we only need to trigger relation_changed if it is a remote app or unit
+
+        if entity == self._model.unit:
```

Comment:
the objects you're comparing here have no comparison methods, so this is the same as saying `entity is self._model.unit`.
It's not clear to me that the fact that the object returned by e.g. `self._model.get_unit` is identical (as in, satisfies `is`) to the `self._model.unit` is guaranteed to be true in the future, or even if it is true by design and isn't just happenstance.

Reply:
I'm happy enough with "if app_or_unit == self._model.unit.name" as we won't have 2 objects with the same name that aren't supposed to be the same.
(done)